src/main/java/edu/kit/compiler/optimizations/unrolling/LoopAnalysis.java

```diff
@@ -83,10 +71,66 @@ public void visitBlock(Block block) {
     }
 
     /**
-     * An exception to indicate that a graph is not reducible. See "Compilers
-     * Principles, Techniques, and Tools" (Aho et al., Chapter 9.6.4).
+     * Returns a map of each block in the graph to its nearest surrounding loop.
+     * More formally, for each block B find a loop L such that B is contained
+     * in the body of L and and there is no Loop L' such that B is contained in
+     * the body of L' and the header of L' is not contained in the body of L.
      */
-    private static final class GraphNotReducibleException extends RuntimeException {
+    private Map<Block, Loop> findSurroundingLoops(Graph graph) {
+        var surroundingLoops = new HashMap<Block, Loop>();
+
+        graph.walkBlocksPostorder(new BlockWalker() {
+            @Override
+            public void visitBlock(Block block) {
+                var loop = findSurroundingLoop(block);
+                if (loop != null) {
+                    assert loop.body.contains(block);
+                    assert loops.values().stream().noneMatch(other -> other.body.contains(block)
+                            && loop.body.contains(other.header));
+                    surroundingLoops.put(block, loop);
+                }
+            }
+        });
+
+        return surroundingLoops;
+    }
+
+    private Loop findSurroundingLoop(Block block) {
+        var dom = block.ptr;
+        while ((dom = binding_irdom.get_Block_idom(dom)) != null) {
+            var loop = loops.get(new Block(dom));
+            if (loop != null && loop.body.contains(block)) {
+                return loop;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * For each loop, checks if it is closed, and invalidates the loop if not.
+     * A loop is said to be closed if there are no jumps from within the loop
+     * body to any block outside of the loop.
+     */
+    private void checkLoopsClosed(Graph graph, Map<Block, Loop> surroundingLoops) {
+        graph.walkBlocks(new BlockWalker() {
+            @Override
+            public void visitBlock(Block block) {
+                var loop = surroundingLoops.get(block);
+                Util.forEachPredBlock(block, (predBlock, i) -> {
+                    var predLoop = surroundingLoops.get(predBlock);
+                    if (predLoop == null) {
+                        return;
+                    }
+                    if (predLoop != null && !predLoop.equals(loop)
```

Comment:
Am I misunderstanding something or do you check if the predecessor of a block is not contained inside another loop that is not its own surrounding loop? How does this handle jumps to other blocks that are not contained in any loops themselves?

Reply:
This checks, whether the control flow edge is a jump from within the body of a loop to a block that is not part of that loop. So, whether `predblock` is inside the body of `predLoop`,  `block` is inside the body of `loop`, `loop` and `predLoop` are not the same loop, and whether the jump is not from body to header.

Also, the second `predLoop != null` is a bit ugly.

Reply:
Okay but what happens if you have a jump to another block that is not contained within a loop? Then`predLoop` should also be null since no predLoop exists right?

Reply:
First of all, `pred` is the jump source, not the destination (reverse control flow dependency). Only the one very specific case is of interest here. I don't need to care about jumps into a loop, or jumps where no loop is involved at all.